### ISSUE TITLE
Compatible with babel 7

### ISSUE THREAD

**Issue by nightire** [ef4/ember-auto-import#119]
Currently when running with babel 7 (through ember-cli-babel v7+), ember-auto-import will throw during build process. It should be easy to fix by upgrade babel-core, just write down a reminder here.

```
=================================================================================

ENV Summary:

  TIME: Thu Aug 30 2018 08:02:56 GMT+0800 (China Standard Time)
  TITLE: ember
  ARGV:
  - /usr/local/bin/node
  - /Users/nightire/Code/very-geek/ember-plyr/node_modules/.bin/ember
  - serve
  EXEC_PATH: /usr/local/bin/node
  TMPDIR: /var/folders/9_/y4f4n_cd5ddbzftt3x0rz19w0000gn/T
  SHELL: /bin/zsh
  PATH:
  - /usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin
  - /Users/nightire/Code/very-geek/ember-plyr/node_modules/.bin
  - /usr/local/bin
  - /usr/bin
  - /bin
  - /usr/sbin
  - /sbin
  - /usr/local/opt/texinfo/bin
  - /usr/local/opt/sqlite/bin
  - /usr/local/opt/openssl/bin
  - /usr/local/opt/libxml2/bin
  - /usr/local/opt/gettext/bin
  - /usr/local/opt/curl/bin
  - /Users/nightire/.cargo/bin
  - /usr/local/opt/gnu-tar/libexec/gnubin
  - /usr/local/Cellar/zplug/2.4.2/bin
  - /usr/local/sbin
  PLATFORM: darwin x64
  FREEMEM: 1405415424
  TOTALMEM: 17179869184
  UPTIME: 601423
  LOADAVG: 3.7802734375,4.15234375,3.4912109375
  CPUS:
  - Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz - 4000
  - Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz - 4000
  - Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz - 4000
  - Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz - 4000
  - Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz - 4000
  - Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz - 4000
  - Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz - 4000
  - Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz - 4000
  ENDIANNESS: LE
  VERSIONS:
  - ares: 1.14.0
  - cldr: 33.1
  - http_parser: 2.8.0
  - icu: 62.1
  - modules: 64
  - napi: 3
  - nghttp2: 1.32.0
  - node: 10.8.0
  - openssl: 1.1.0h
  - tz: 2018e
  - unicode: 11.0
  - uv: 1.22.0
  - v8: 6.7.288.49-node.19
  - zlib: 1.2.11

ERROR Summary:

  - broccoliBuilderErrorStack: [undefined]
  - codeFrame: [undefined]
  - errorMessage: Requires Babel "^7.0.0-0", but was loaded with "6.26.3". If you are sure you have a compatible version of @babel/core, it is likely that something in your build process is loading the wrong version. Inspect the stack trace of this error to look for the first entry that doesn't mention "@babel/core" or "babel-core" to see what is calling Babel.
  - errorType: [undefined]
  - location:
    - column: [undefined]
    - file: [undefined]
    - line: [undefined]
  - message: Requires Babel "^7.0.0-0", but was loaded with "6.26.3". If you are sure you have a compatible version of @babel/core, it is likely that something in your build process is loading the wrong version. Inspect the stack trace of this error to look for the first entry that doesn't mention "@babel/core" or "babel-core" to see what is calling Babel.
  - name: Error
  - nodeAnnotation: [undefined]
  - nodeName: [undefined]
  - originalErrorMessage: [undefined]
  - stack: Error: Requires Babel "^7.0.0-0", but was loaded with "6.26.3". If you are sure you have a compatible version of @babel/core, it is likely that something in your build process is loading the wrong version. Inspect the stack trace of this error to look for the first entry that doesn't mention "@babel/core" or "babel-core" to see what is calling Babel.
    at throwVersionError (/Users/nightire/Code/very-geek/ember-plyr/node_modules/@babel/helper-plugin-utils/lib/index.js:65:11)
    at Object.assertVersion (/Users/nightire/Code/very-geek/ember-plyr/node_modules/@babel/helper-plugin-utils/lib/index.js:13:11)
    at _default (/Users/nightire/Code/very-geek/ember-plyr/node_modules/@babel/plugin-proposal-decorators/lib/index.js:35:7)
    at /Users/nightire/Code/very-geek/ember-plyr/node_modules/@babel/helper-plugin-utils/lib/index.js:19:12
    at Function.memoisePluginContainer (/Users/nightire/Code/very-geek/ember-plyr/node_modules/babel-core/lib/transformation/file/options/option-manager.js:113:13)
    at Function.normalisePlugin (/Users/nightire/Code/very-geek/ember-plyr/node_modules/babel-core/lib/transformation/file/options/option-manager.js:146:32)
    at /Users/nightire/Code/very-geek/ember-plyr/node_modules/babel-core/lib/transformation/file/options/option-manager.js:184:30
    at Array.map (<anonymous>)
    at Function.normalisePlugins (/Users/nightire/Code/very-geek/ember-plyr/node_modules/babel-core/lib/transformation/file/options/option-manager.js:158:20)
    at OptionManager.mergeOptions (/Users/nightire/Code/very-geek/ember-plyr/node_modules/babel-core/lib/transformation/file/options/option-manager.js:234:36)
    at OptionManager.init (/Users/nightire/Code/very-geek/ember-plyr/node_modules/babel-core/lib/transformation/file/options/option-manager.js:368:12)
    at File.initOptions (/Users/nightire/Code/very-geek/ember-plyr/node_modules/babel-core/lib/transformation/file/index.js:212:65)
    at new File (/Users/nightire/Code/very-geek/ember-plyr/node_modules/babel-core/lib/transformation/file/index.js:135:24)
    at Analyzer.buildParserOptions (/Users/nightire/Code/very-geek/ember-plyr/node_modules/ember-auto-import/js/analyzer.js:50:17)
    at new Analyzer (/Users/nightire/Code/very-geek/ember-plyr/node_modules/ember-auto-import/js/analyzer.js:39:35)
    at AutoImport.analyze (/Users/nightire/Code/very-geek/ember-plyr/node_modules/ember-auto-import/js/auto-import.js:41:24)
    at Object.toTree (/Users/nightire/Code/very-geek/ember-plyr/node_modules/ember-auto-import/js/index.js:21:59)
    at /Users/nightire/Code/very-geek/ember-plyr/node_modules/ember-cli-preprocess-registry/preprocessors.js:180:26
    at Array.forEach (<anonymous>)
    at processPlugins (/Users/nightire/Code/very-geek/ember-plyr/node_modules/ember-cli-preprocess-registry/preprocessors.js:178:11)
    at Function.module.exports.preprocessJs (/Users/nightire/Code/very-geek/ember-plyr/node_modules/ember-cli-preprocess-registry/preprocessors.js:171:10)
    at Class.preprocessJs (/Users/nightire/Code/very-geek/ember-plyr/node_modules/ember-cli/lib/models/addon.js:1275:25)
    at Class.processedAddonJsFiles (/Users/nightire/Code/very-geek/ember-plyr/node_modules/ember-cli/lib/models/addon.js:1289:33)
    at Class.compileAddon (/Users/nightire/Code/very-geek/ember-plyr/node_modules/ember-cli/lib/models/addon.js:1170:24)
    at Class.treeForAddon (/Users/nightire/Code/very-geek/ember-plyr/node_modules/ember-cli/lib/models/addon.js:826:26)

=================================================================================
```

**Comment by ef4** [comment]
Thanks. Yeah, I knew this was coming. 

It’s not actually easy to fix, because a single copy of ember-auto-import handles the app and all the addons, so we need to become compatible with both Babel 6 and Babel 7.

**Comment by ef4** [comment]
The architectural issue here is that we use the babel config from each app or addon to instantiate a babel parser and use that parser to examine that app or addon's source code for imports. If one of them is using babel 7, their babel config won't work with the babel 6 parser.

We could just add both parsers as dependencies of ember-auto-import. But it would be even better to move our analysis phase into a babel plugin. That way we could just include two babel plugins, one for 6 and one for 7. This also results in less total work, since you don't have to re-parse code that was going to be parsed by ember-cli-babel anyway.

The only significant challenge to switching to an analysis babel plugin is making sure we have as much caching as ember-cli-babel has. We will only find imports via side-effect of babel transpilation, so if a file is cached by ember-cli-babel, we won't see it, so we'll need to make sure we have its imports cached too.

**Comment by ef4** [comment]
I don't think it does but feel free to experiment.

**Comment by josemarluedke** [comment]
Any needs to allowing apps to use babel 7 with ember-auto-import?

**Comment by buschtoens** [comment]
@josemarluedke As written in https://github.com/ef4/ember-auto-import/issues/119#issuecomment-417270890, the "easy" solution is:

> The architectural issue here is that we use the babel config from each app or addon to instantiate a babel parser and use that parser to examine that app or addon's source code for imports. If one of them is using babel 7, their babel config won't work with the babel 6 parser.
>
> We could just add both parsers as dependencies of ember-auto-import.

The _proper_ solution is:

> But it would be even better to move our analysis phase into a babel plugin.

Since my company is low-key interested in moving https://github.com/bgentry/ember-apollo-client/pull/175 forward, I might be able to allocate some time to do the "easy" solution @dfreeman has suggested in https://github.com/bgentry/ember-apollo-client/pull/175#issuecomment-436298320:

> Maybe separating the Babel 7 support from that work and doing a first pass that continues with the status quo but can work with either the Babel 6 or 7 parser would be a good first step?

But let me not stop you from giving it a go yourself or even attempting the proper solution. 😄

**Comment by ef4** [comment]
I’m totally fine with separating these goals and doing the simpler Babel 7 integration. The only downside is more dependencies for auto-import itself.

**Comment by buschtoens** [comment]
> Is it possible to (easily) do a Babel-7-only version and just bump the major version of the plugin until a solution that works for both 6 and 7 is implemented?

Releasing a Babel 7 only version doesn't make sense IMO, because in any case, a new parser handler that uses Babel 7 needs to be written. Keeping the old Babel 6 parser in parallel probably does not mean a lot of extra work.

**Comment by ef4** [comment]
Fixed in #153 and released as v1.2.16.
